_posts/2021-03-05-automatically-generated-api-tests-with-clojure-and-reitit.md

```diff
@@ -0,0 +1,332 @@
+test which walks over the entire API.
+
+My current customer project is using something along these lines to test that
+the entire API handles credentials and permissions appropriately and doesn't
+randomly crash at some endpoints. As a consequence, the tests generated by
+these methods happen to create an almost complete smoke test suite, where a new
+endpoint can be added with a few lines of code. It also forces our API to behave
+consistently and give out similar errors independent of the endpoint.
+
+Example API
+===========
+
+For a small example api, we have the following endpoints:
+- `POST:/api/resource` - Create a new resource
+- `GET:/api/resource/:resource-id` - Get the created resource by ID
+- `PUT:/api/resource/:resource-id` - Update the resource
+- `POST:/api/authorization` - Add a user to a resource
+- `DELETE:/api/authorization` - Delete a user from a resource
+
+Only users added to a resource can update it. The creator of the resource is
+added when creating the resource.
+
+The endpoints behave in the following way: All requests without the Auth header
+return an error. Creating resources returns the created resource id. The id can
+be used to find the created resource.
+
+```
+$ curl -d "Hello" localhost:3001/api/resource   
+{:cause :missing-authentication}
+$ curl -H "Auth: 1" -H "Content-Type: application/json" -d '{"content": "Hello"}' localhost:3001/api/resource
+{:resource-id "4"}
+$ curl -H "Auth: 1" localhost:3001/api/resource/4
+{:content "Hello"}
+```
+
+Using a different Auth id than the one used to create the resource initially
+returns an error, but if the original user is used to add the other user into
+the resource, the other user can also access the resource.
+
+```
+$ curl -H "Auth: 2" localhost:3001/api/resource/4
+{:cause :not-authorized}
+$ curl -H "Auth: 1" -H "Content-Type: application/json" -d '{"resource-id": "4", "token": "2"}' localhost:3001/api/authorization
+{}%                                 
+$ curl -H "Auth: 2" localhost:3001/api/resource/4
+{:content "Hello"}
+```
+
+The API looks like the following piece of code:
+
+```clojure
+(def api
+  [["/api"
+   ["/authorization"
+    {:post   add-authorization
+     :delete remove-authorization}]
+   ["/resource"
+    ["" {:post create-resource}]
+    ["/:resource-id"
+     {:get    read-resource
+      :put    update-resource
+      :delete delete-resource}]]]])
+```
+
+Testing
+=======
+
+Traditionally testing an endpoint like this is relatively simple. Just make some requests and
+check that the returned value is the expected one.
+
+```clojure
+(defn- request
+  ([method token uri]
+   (request method token uri nil))
+  ([method token uri body]
+   (-> (handler/app
+         {:uri            uri
+          :request-method method
+          :headers        {"auth" token}
+          :body-params    body})
+       (select-keys [:status :body])
+       (update :body read-string))))
```

Comment:
I know it doesn't necessarily matter here because it's just test code, but `read-string` can execute code. :P

```clojure
(update {:body "#=(clojure.java.shell/sh \"echo\" \"hello\" )"} :body read-string)
;;=> {:body {:exit 0, :out "hello\", :err ""}}
```

Maybe `clojure.edn/read-string`? Not a big deal, though.